platform/engineering/design-docs/2020-04-16-integration-testing-with-consumer-driven-contract-tests.md

```diff
@@ -0,0 +1,304 @@
+
+There has been at least one incident that revealed the ineffectiveness of the current test strategy in detecting breaking changes from the API. In this case, an API endpoint was modified independently of the form that invoked it, causing the form to break. Tests did not catch this as they were assuming an outdated expected response.
+
+To address the gap in integration testing, the VSP FE Tools team has done some discovery on **consumer-driven contract testing (CDCT)** as a method of integration testing and [evaluated Pact as a CDCT solution](https://github.com/department-of-veterans-affairs/va.gov-team/blob/3a8f4953b5e77edcbd39fd3db073232cdaca0ea0/teams/vsp/teams/tools/frontend/2019-11-13-consumer-driven-contract-testing-with-pact.md).
+
+We also conducted a [testing tools audit](https://github.com/department-of-veterans-affairs/va.gov-team/blob/fa7b02be2d7a71821a893a55e61a26ba53f0b770/teams/vsp/teams/tools/frontend/2020-03-31-testing-tools-audit.md#integration-testing) that reiterated the lack of integration testing in our current processes. As part of that audit, Pact is still under consideration as our preferred CDCT solution.
+
+### High Level Design
+
+The design of the integration testing solution will be described in terms of using [Pact, a consumer-driven contract testing framework](https://docs.pact.io/).
+
+This [animated diagram from Pactflow (starting at slide 3)](https://pactflow.io/how-pact-works/#slide-3) illustrates how Pact works.
+
+#### [What Is Contract Testing?](https://docs.pact.io/#what-is-contract-testing)
+
+> **Contract testing is a technique for testing an integration point by checking each application in isolation to ensure the messages it sends or receives conform to a shared understanding that is documented in a "contract".**
+
+> For applications that communicate via HTTP, these "messages" would be the HTTP request and response, and for an application that used queues, this would be the message that goes on the queue.
+
+The contract involves two parties, the **consumer** and **provider**. For our purposes, the FE application is the consumer consuming an endpoint from the provider, which is the API.
+
+In a consumer-driven context, the consumer generates the contract and the provider verifies it.
+
+The process of using Pact can be broken into multiple steps:
+1. Consumers run unit tests that test their request and response interactions with the provider's endpoints.
+2. The tests generate contracts, also referred to as **pacts** in the Pact framework.
+3. The pacts are published to a central location or a broker to be versioned and shared with the provider.
+4. The provider verifies the pacts against an actual running instance of itself and validates that the actual response matches the expected response.
+5. Results of the verification can be published back to the broker if there is a broker involved.
+6. Builds will depend on successful verification results to pass.
+
+## Specifics
+
+### Detailed Design
```

Comment:
reading this (esp. steps 2-6) feels like it would benefit from some sort of visual diagram (sequence diagram maybe?) showing the control flow: CircleCI, vets-website, pact broker, vets-api, and when each type of test is run (especially since parts of it are async).